drivers/can/can_mcan.c

```diff
@@ -0,0 +1,875 @@
+		can->cccr &= ~(CAN_MCAN_CCCR_TEST | CAN_MCAN_CCCR_MON);
+		break;
+
+	case CAN_SILENT_MODE:
+		LOG_DBG("Config silent mode");
+		can->cccr &= ~CAN_MCAN_CCCR_TEST;
+		can->cccr |= CAN_MCAN_CCCR_MON;
+		break;
+
+	case CAN_LOOPBACK_MODE:
+		LOG_DBG("Config loopback mode");
+		can->cccr &= ~CAN_MCAN_CCCR_MON;
+		can->cccr |= CAN_MCAN_CCCR_TEST;
+		can->test |= CAN_MCAN_TEST_LBCK;
+		break;
+
+	case CAN_SILENT_LOOPBACK_MODE:
+		LOG_DBG("Config silent loopback mode");
+		can->cccr |= (CAN_MCAN_CCCR_TEST | CAN_MCAN_CCCR_MON);
+		can->test |= CAN_MCAN_TEST_LBCK;
+		break;
+	default:
+		break;
+	}
+
+	ret = can_leave_init_mode(can, K_MSEC(CAN_INIT_TIMEOUT));
+	if (ret) {
+		LOG_ERR("Failed to leave init mode");
+	}
+
+	return 0;
+}
+
+int can_mcan_init(const struct device *dev, const struct can_mcan_config *cfg,
+		  struct can_mcan_msg_sram *msg_ram,
+		  struct can_mcan_data *data)
+{
+	struct can_mcan_reg  *can = cfg->can;
+	struct can_timing timing;
+#ifdef CONFIG_CAN_FD_MODE
+	struct can_timing timing_data;
+#endif
+	int ret;
+
+	k_mutex_init(&data->inst_mutex);
+	k_mutex_init(&data->tx_mtx);
+	k_sem_init(&data->tx_sem, NUM_TX_BUF_ELEMENTS, NUM_TX_BUF_ELEMENTS);
+	for (int i = 0; i < ARRAY_SIZE(data->tx_fin_sem); ++i) {
+		k_sem_init(&data->tx_fin_sem[i], 0, 1);
+	}
+
+	ret = can_exit_sleep_mode(can);
+	if (ret) {
+		LOG_ERR("Failed to exit sleep mode");
+		return -EIO;
+	}
+
+	ret = can_enter_init_mode(can, K_MSEC(CAN_INIT_TIMEOUT));
+	if (ret) {
+		LOG_ERR("Failed to enter init mode");
+		return -EIO;
+	}
+
+	/* Configuration Change Enable */
+	can->cccr |= CAN_MCAN_CCCR_CCE;
+
+	LOG_DBG("IP rel: %lu.%lu.%lu %02lu.%lu.%lu",
+		(can->crel & CAN_MCAN_CREL_REL) >> CAN_MCAN_CREL_REL_POS,
+		(can->crel & CAN_MCAN_CREL_STEP) >> CAN_MCAN_CREL_STEP_POS,
+		(can->crel & CAN_MCAN_CREL_SUBSTEP) >>
+			     CAN_MCAN_CREL_SUBSTEP_POS,
+		(can->crel & CAN_MCAN_CREL_YEAR) >> CAN_MCAN_CREL_YEAR_POS,
+		(can->crel & CAN_MCAN_CREL_MON) >> CAN_MCAN_CREL_MON_POS,
+		(can->crel & CAN_MCAN_CREL_DAY) >> CAN_MCAN_CREL_DAY_POS);
+
+#ifndef CONFIG_CAN_STM32FD
+	can->sidfc = ((uint32_t)msg_ram->std_filt & CAN_MCAN_SIDFC_FLSSA_MSK) |
+		     (ARRAY_SIZE(msg_ram->std_filt) << CAN_MCAN_SIDFC_LSS_POS);
+	can->xidfc = ((uint32_t)msg_ram->ext_filt & CAN_MCAN_XIDFC_FLESA_MSK) |
+		     (ARRAY_SIZE(msg_ram->ext_filt) << CAN_MCAN_XIDFC_LSS_POS);
+	can->rxf0c = ((uint32_t)msg_ram->rx_fifo0 & CAN_MCAN_RXF0C_F0SA) |
+		     (ARRAY_SIZE(msg_ram->rx_fifo0) << CAN_MCAN_RXF0C_F0S_POS);
+	can->rxf1c = ((uint32_t)msg_ram->rx_fifo1 & CAN_MCAN_RXF1C_F1SA) |
+		     (ARRAY_SIZE(msg_ram->rx_fifo1) << CAN_MCAN_RXF1C_F1S_POS);
+	can->rxbc = ((uint32_t)msg_ram->rx_buffer & CAN_MCAN_RXBC_RBSA);
+	can->txefc = ((uint32_t)msg_ram->tx_event_fifo & CAN_MCAN_TXEFC_EFSA_MSK) |
+		     (ARRAY_SIZE(msg_ram->tx_event_fifo) <<
+		     CAN_MCAN_TXEFC_EFS_POS);
+	can->txbc = ((uint32_t)msg_ram->tx_fifo & CAN_MCAN_TXBC_TBSA) |
+		    (ARRAY_SIZE(msg_ram->tx_fifo) << CAN_MCAN_TXBC_TFQS_POS);
```

Comment:
This doesn't look correct to me. The start addresses should be relative to the `msg_ram` and not as the absolute address in the MCU's RAM. In addition to that, it has to be left-shifted by 2 bits for most of the registers.

Reply:
Never mind, I understand now that the addresses are actually absolute addresses, but they the message RAM must be located in the first 64k so that a 16-bit address is sufficient.